### PR TITLE
2.4.0 migration script and upgrade tests

### LINKED ISSUE #848: 2.4.0 upgrade tests + migration script
Before upgrading from v2.3.0 to v2.4.0 we need to make sure the upgrade does not corrupt the protocol.

Current diff: https://github.com/bosonprotocol/boson-protocol-contracts/compare/v2.3.0..main

## Breaking changes
- [x] setRoyaltyPercentage/getRoyaltyPercentage does not work anymore
- [x] all "createOffer" methods accept a new parameter _feeLimit
- [x] releasing funds for existing exchanges and offers should work normally

## New methods
- [x] setMaxRoyaltyPercentage (replaces setMaxRoyaltyPecentage)
- [x] getEIP2981Royalties
- [x] getRoyalties
- [x] addRoyaltyRecipients
- [x] updateRoyaltyRecipients
- [x] removeRoyaltyRecipients
- [x] getSellersCollectionsPaginated
- [x] getRoyaltyRecipients
- [x] updateOfferRoyaltyRecipients
- [x] updateOfferRoyaltyRecipientsBatch
- [x] commitToPriceDiscoveryOffer
- [x] sequentialCommitToOffer
- [x] royaltyInfo (not really a new method, but logic was moved from voucher to the protocol and should still work)
- [x] onERC721Received

## Bug fixes
- [x] FundsEncumbered emitted during escalate dispute

## New storage slots without getters
- [x] ProtocolLookups.royaltyRecipientIndexBySellerAndRecipient
- [x] ProtocolLookups.royaltyRecipientIdByWallet
- [x] ProtocolEntities.exchangeCosts
- [x] ProtocolEntities.royaltyRecipients
- [x] ProtocolStatus.incomingVoucherId
- [x] ProtocolStatus.incomingVoucherCloneAddress

## Migration script ToDo
- [x] Pause the seller and offer facet
- [x] Get info about the existing royalty recipients
- [x] Prepare data to backfill the allowed royalty recipients
- [x] handle `initV2_4_0External`
- [x] add/remove correct metatx method
- [x] Unpause everything

### PR DESCRIPTION
**Pull request by zajck**
Close #848 

- New migration script, which includes data backfilling (for royalty recipients) and deployment of price discovery client
- New upgrade tests (v2.3.0-v2.4.0)
- Minor fixes and changes in existing scripts

### CODE REVIEW

**Review by albertfolch-redeemeum — DISMISSED** [review]

**Review by levalleux-ludo — APPROVED** [review]
Approved with 1 comment.

File: scripts/migrations/migrate_2.4.0.js [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+  const offerHandler = await getContractAt("IBosonOfferHandler", protocolAddress);
+  const nextOfferId = await offerHandler.getNextOfferId();
+
+  for (let i = 1n; i < nextOfferId; i++) {
+    const [, offer, offerDates] = await offerHandler.getOffer(i);
+
+    // if voided or expired, skip
+    // const currentTime = BigInt(Math.floor(Date.now() / 1000));
+    const currentTime = 0n;
```
[the remark this answers is not in the document]
Comment (zajck) [reply scripts/migrations/migrate_2.4.0.js]: removed

File: scripts/migrations/migrate_2.4.0.js [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+  if (params) {
+    if (params.WrappedNative) {
+      console.log("Using WrappedNative from params");
+      config.facetsToInit.PriceDiscoveryHandlerFacet.constructorArgs[0] = params.WrappedNative;
+      config.facetsToInit.SequentialCommitHandlerFacet.constructorArgs[0] = params.WrappedNative;
+    }
+  }
+
+  try {
+    if (env != "upgrade-test") {
```
Comment (albertfolch-redeemeum) [inline scripts/migrations/migrate_2.4.0.js:75]:
```suggestion
    if (env !== "upgrade-test") {
```

File: scripts/migrations/migrate_2.4.0.js [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+      if (statusOutput.stdout) {
+        throw new Error("Local changes found. Please stash them before upgrading");
+      }
+    }
+
+    const { chainId } = await ethers.provider.getNetwork();
+    const contractsFile = readContracts(chainId, network, env);
+
+    if (contractsFile?.protocolVersion != "2.3.0") {
```
Comment (albertfolch-redeemeum) [inline scripts/migrations/migrate_2.4.0.js:88]:
```suggestion
    if (contractsFile?.protocolVersion !== "2.3.0") {
```

File: scripts/migrations/migrate_2.4.0.js [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+    let contracts = contractsFile?.contracts;
+
+    // Get addresses of currently deployed contracts
+    const accessControllerAddress = contracts.find((c) => c.name === "AccessController")?.address;
+
+    const accessController = await getContractAt("AccessController", accessControllerAddress);
+
+    const signer = (await getSigners())[0].address;
+    if (env == "upgrade-test") {
```
Comment (albertfolch-redeemeum) [inline scripts/migrations/migrate_2.4.0.js:100]:
```suggestion
    if (env === "upgrade-test") {
```

File: scripts/migrations/migrate_2.4.0.js [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+      // Grant PAUSER role to the deployer
+      await accessController.grantRole(Role.PAUSER, signer);
+    } else {
+      checkRole(contracts, "PAUSER", signer);
+    }
+
+    const protocolAddress = contracts.find((c) => c.name === "ProtocolDiamond")?.address;
+
+    if (env != "upgrade-test") {
```
Comment (albertfolch-redeemeum) [inline scripts/migrations/migrate_2.4.0.js:109]:
```suggestion
    if (env !== "upgrade-test") {
```

**Review by albertfolch-redeemeum — DISMISSED** [review]

**Review by levalleux-ludo — APPROVED** [review]
LGTM

**Review by albertfolch-redeemeum — APPROVED** [review]

**Pull request merged**